File: task_manager.go
package main

import (
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"syscall"
)

type Task struct {
	Name             string `json:"name"`
	Command          string `json:"command"`
	WorkingDirectory string `json:"working_directory"`
	Group            string `json:"group,omitempty"`
}

type TaskManager struct {
	configPath string
	logDir     string
	Tasks      []*Task
	processes  map[*Task]*exec.Cmd
}

func NewTaskManager() *TaskManager {
	home, _ := os.UserHomeDir()
	tm := &TaskManager{
		configPath: filepath.Join(home, ".config", "devtray", "config.json"),
		logDir:     filepath.Join(home, ".cache", "devtray", "logs"),
		Tasks:      []*Task{},
		processes:  make(map[*Task]*exec.Cmd),
	}
	tm.LoadConfig()
	return tm
}

func (tm *TaskManager) LoadConfig() {
	data, err := os.ReadFile(tm.configPath)
	if err != nil {
		tm.Tasks = []*Task{}
		return
	}
	var tasks []*Task
	err = json.Unmarshal(data, &tasks)
	if err != nil {
		tm.Tasks = []*Task{}
		return
	}
	tm.Tasks = tasks
}

func (tm *TaskManager) SaveConfig() {
	os.MkdirAll(filepath.Dir(tm.configPath), 0755)
	data, _ := json.MarshalIndent(tm.Tasks, "", "  ")
	os.WriteFile(tm.configPath, data, 0644)
}

func (tm *TaskManager) AddTask(task *Task) {
	tm.Tasks = append(tm.Tasks, task)
	tm.SaveConfig()
}

func (tm *TaskManager) RemoveTask(task *Task) {
	for i, t := range tm.Tasks {
		if t == task {
			tm.StopTask(task)
			tm.Tasks = append(tm.Tasks[:i], tm.Tasks[i+1:]...)
			tm.SaveConfig()
			return
		}
	}
}

func (tm *TaskManager) UpdateTask(oldTask *Task, newTask *Task) {
	for i, t := range tm.Tasks {
		if t == oldTask {
			wasRunning := tm.IsRunning(oldTask)
			if wasRunning {
				tm.StopTask(oldTask)
			}
			tm.Tasks[i] = newTask
			tm.SaveConfig()
			if wasRunning {
				tm.StartTask(newTask)
			}
			return
		}
	}
}

func (tm *TaskManager) StartTask(task *Task) {
	if tm.IsRunning(task) {
		return
	}
	os.MkdirAll(tm.logDir, 0755)
	logFile := filepath.Join(tm.logDir, task.Name+".log")
	f, _ := os.OpenFile(logFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)

	cwd := task.WorkingDirectory
	if cwd == "" {
		cwd, _ = os.Getwd()
	} else {
		cwd = os.ExpandEnv(cwd)
		if len(cwd) > 0 && cwd[0] == '~' {
			home, _ := os.UserHomeDir()
			if len(cwd) == 1 {
				cwd = home
			} else if cwd[1] == '/' {
				cwd = filepath.Join(home, cwd[2:])
			}
		}
	}

	cmd := exec.Command("sh", "-c", task.Command)
	cmd.Dir = cwd
	cmd.Stdout = f
	cmd.Stderr = f
	// Create new process group
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}

	err := cmd.Start()
	if err == nil {
		tm.processes[task] = cmd
		// Wait in background to reap
		go func() {
			cmd.Wait()
			f.Close()
		}()
	} else {
		f.Close()
	}
}

func (tm *TaskManager) StopTask(task *Task) {
	if cmd, ok := tm.processes[task]; ok {
		if cmd.Process != nil {
			// Kill process group forcefully
			syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
		}
		delete(tm.processes, task)
	}
}

func (tm *TaskManager) IsRunning(task *Task) bool {
	if cmd, ok := tm.processes[task]; ok {
		// ProcessState is set when Wait() completes
		if cmd.ProcessState != nil && cmd.ProcessState.Exited() {
			delete(tm.processes, task)
			return false
		}
		// If Process is set but ProcessState is nil, it's running
		if cmd.Process != nil && cmd.ProcessState == nil {
			return true
		}
	}
	return false
}

func (tm *TaskManager) StopAll() {
	for _, t := range tm.Tasks {
		tm.StopTask(t)
	}
}

File: src/devtray/task_manager.py
import os
import json
import subprocess
from pathlib import Path

class Task:
    def __init__(self, name: str, command: str, working_directory: str):
        self.name = name
        self.command = command
        self.working_directory = working_directory
        
    def to_dict(self):
        return {
            "name": self.name,
            "command": self.command,
            "working_directory": self.working_directory
        }
        
    @classmethod
    def from_dict(cls, data: dict):
        return cls(
            name=data.get("name", "Unknown"),
            command=data.get("command", ""),
            working_directory=data.get("working_directory", os.getcwd())
        )

class TaskManager:
    def __init__(self, config_path: str = None, log_dir: str = None):
        self.config_path = Path(config_path) if config_path else Path.home() / ".config" / "devtray" / "config.json"
        self.log_dir = Path(log_dir) if log_dir else Path.home() / ".cache" / "devtray" / "logs"
        self.tasks = []
        self._processes = {} # map Task object id to (subprocess.Popen, log_file_handle)
        
        self.load_config()

    def load_config(self):
        if not self.config_path.exists():
            self.tasks = []
            return
            
        try:
            with open(self.config_path, "r", encoding="utf-8") as f:
                data = json.load(f)
                self.tasks = [Task.from_dict(t) for t in data]
        except (json.JSONDecodeError, FileNotFoundError):
            self.tasks = []
            
    def save_config(self):
        self.config_path.parent.mkdir(parents=True, exist_ok=True)
        data = [t.to_dict() for t in self.tasks]
        with open(self.config_path, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2)

    def get_tasks(self):
        return self.tasks

    def add_task(self, task: Task):
        self.tasks.append(task)
        
    def remove_task(self, task: Task):
        if task in self.tasks:
            self.stop_task(task)
            self.tasks.remove(task)

    def start_task(self, task: Task):
        if self.is_running(task):
            return # already running
            
        self.log_dir.mkdir(parents=True, exist_ok=True)
        log_file_path = self.log_dir / f"{task.name}.log"
        
        # Open log file in append mode (or write mode)
        log_file = open(log_file_path, "w", encoding="utf-8")
        
        cwd = task.working_directory
        if not os.path.exists(cwd):
            cwd = os.getcwd()
            
        process = subprocess.Popen(
            task.command,
            shell=True,
            cwd=cwd,
            stdout=log_file,
            stderr=subprocess.STDOUT
        )
        
        self._processes[id(task)] = (process, log_file)
        
    def stop_task(self, task: Task):
        if id(task) in self._processes:
            process, log_file = self._processes[id(task)]
            process.terminate() # Send SIGTERM
            
            try:
                process.wait(timeout=3)
            except subprocess.TimeoutExpired:
                process.kill() # Force kill if stubborn
                
            log_file.close()
            del self._processes[id(task)]
            
    def is_running(self, task: Task) -> bool:
        if id(task) in self._processes:
            process, _ = self._processes[id(task)]
            if process.poll() is None:
                return True
            else:
                # Process has finished, clean up state
                _, log_file = self._processes[id(task)]
                log_file.close()
                del self._processes[id(task)]
                return False
        return False
        
    def stop_all(self):
        for task in self.tasks:
            self.stop_task(task)
